kernel/sched.c

```diff
@@ -681,6 +685,26 @@ void z_sched_start(struct k_thread *thread)
 	z_reschedule(&sched_spinlock, key);
 }
 
+#ifdef CONFIG_SMP
+/**
+ * @brief Test if the CPU is deadlocked waiting on each other in ISRs
+ *
+ * Test if the CPUs are deadlocked waiting on each other in ISRs.
+ *
+ * @return true if deadlock detected, false if not
+ */
+static bool is_cpu_isr_deadlocked(struct _cpu *cpu)
+{
+	unsigned int  i;
+
+	for (i = 0; (i < CONFIG_MP_MAX_NUM_CPUS) && (cpu != NULL); i++) {
+		cpu = cpu->spinning_on;
+	}
```

Comment:
I would normally expect a kind of circular dependency for a deadlock scenario. E.g. if any traversed CPU node points back to the original `cpu`, then there is a deadlock.

Is there a key point that I'm missing here?

Reply:
I know that classic method of finding a loop in an unknown number of linked nodes is to have two pointers, one advancing twice as fast as the other, and if they match before the end is reached, a loop has been detected. However, what we have here is a known maximum number of nodes (and a relatively small number of them at that).

If we have traversed CONFIG_MP_MAX_NUM_CPUS nodes (one for each CPU in the system), and we still have not reached NULL, then we know that there is a loop--either it is a loop that consists of all the CPUs, or a subset of them. If we reach NULL, then we know that our CPU node is not part of any loop.